### PR TITLE
[now-build-utils][now-static-build][now-cli] Update detectors and undo previous changes

### PR DESCRIPTION
**Pull request by AndyBitz**
Update detectors and undo previous changes.

Instead of using the detectors, we'll use `@now/frameworks` to only detect the framework, `@now/static-build` will take care of finding the dev and build command, since we'll have all files available there.

[PRODUCT-781]
[PRODUCT-783]
[PRODUCT-750]

[PRODUCT-781]: https://zeit.atlassian.net/browse/PRODUCT-781
[PRODUCT-783]: https://zeit.atlassian.net/browse/PRODUCT-783
[PRODUCT-750]: https://zeit.atlassian.net/browse/PRODUCT-750

### CODE REVIEW

File: packages/frameworks/frameworks.json [hunk 1]
```diff
@@ -4,70 +4,329 @@
     "slug": "nextjs",
     "logo": "https://raw.githubusercontent.com/zeit/now/master/packages/frameworks/logos/next.svg",
     "tagline": "Next.js makes you productive with React instantly — whether you want to build static or dynamic sites. ",
-    "website": "https://nextjs.org"
+    "website": "https://nextjs.org",
+    "detectors": {
+      "every": [
+        {
+          "file": "package.json",
+          "matchContent": "\"(dev)?(d|D)ependencies\":\\s*{[^}]*\"next\":\\s*\".+?\"[^}]*}"
```
Comment (styfle) [inline packages/frameworks/frameworks.json:12]: This seems very verbose and prone to error.

Is there anyway we could minimize the amount of copy/pasta for each framework?

File: packages/frameworks/frameworks.json [hunk 2]
```diff
@@ -4,70 +4,329 @@
...
+      "every": [
+        {
+          "file": "package.json",
+          "matchContent": "\"(dev)?(d|D)ependencies\":\\s*{[^}]*\"@angular\\/cli\":\\s*\".+?\"[^}]*}"
+        }
+      ]
+    }
   },
   {
-    "name": "Hugo",
```
Comment (styfle) [inline packages/frameworks/frameworks.json:52]: I thought Hugo was deleted but its just moved. How did you choose the order?

Comment (AndyBitz) [reply packages/frameworks/frameworks.json:52]: I've sorted them after `frameworks.ts` in `@now/static-build`, to make sure they are detected in the same order

File: packages/frameworks/frameworks.json (hunk 1 above)
Comment (AndyBitz) [reply packages/frameworks/frameworks.json:12]: We tried to make it as generic as possible

cc @leo

File: packages/frameworks/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  name: string;
+  slug: string;
+  logo: string;
+  tagline: string;
+  website: string;
+  detectors?: {
+    every?: FrameworkDetectionItem[];
+    some?: FrameworkDetectionItem[];
+  };
+}
```
Comment (styfle) [inline packages/frameworks/index.d.ts:16]: Is this being type checked at build time or are you manually matching this to the json file?

If its manual, can we rename the `frameworks.json` to `frameworks.ts` and generate this `.d.ts` file?

File: packages/now-build-utils/src/fs/node-version.ts [hunk 4]
```diff
@@ -17,9 +17,9 @@ const allOptions: NodeVersion[] = [
 const supportedOptions = allOptions.filter(o => !isDiscontinued(o));
 
 // This version should match Fargate's default in the PATH
-// Today that is Node 8
+// Today that is Node 12
 export const defaultSelection = supportedOptions.find(
-  o => o.major === 8
+  o => o.major === 12
```
Comment (styfle) [inline packages/now-build-utils/src/fs/node-version.ts]: Why is the node version being changed in this PR?

This should be a separate PR

Comment (AndyBitz) [reply packages/now-build-utils/src/fs/node-version.ts]: Will undo, this was just so the tests wouldn't print the warning

File: packages/frameworks/index.d.ts (hunk 3 above)
Comment (AndyBitz) [reply packages/frameworks/index.d.ts:16]: I agree that the `framework.json` file should come from a .ts file, but I think we need a `.json` file in the repo, at least it looks like it's using the repo url for the `logo` data.

In any case I'd like to update them in a different PR, they are only used for tests at the moment

File: packages/now-build-utils/src/detectors/filesystem.ts [hunk 5]
```diff
@@ -58,84 +52,4 @@ export default abstract class DetectorFilesystem {
...
-
-  public readJsonOrNull = async <T>(name: string): Promise<T | null> => {
-    return nullEnoent(this.readJson<T>(name));
-  };
-
-  public readPackageJson = async (): Promise<PackageJson | null> => {
-    return await this.readJsonOrNull<PackageJson>('package.json');
-  };
-
-  public readConfigFile = async <T>(...names: string[]): Promise<T | null> => {
```
Comment (styfle) [inline packages/now-build-utils/src/detectors/filesystem.ts:83]: Any reason why this was moved out of the DetectorFileSystem class?

Seems like we should either use it here or delete the whole file 🤔 

cc @TooTallNate

Comment (AndyBitz) [reply packages/now-build-utils/src/detectors/filesystem.ts:83]: @styfle yes, because it's now used in `@now/static-build` instead, and `DetectorFileSystem` doesn't require it anymore

File: packages/now-build-utils/tsconfig.json [hunk 6]
```diff
@@ -13,7 +13,8 @@
     "outDir": "./dist",
     "types": ["node", "jest"],
     "strict": true,
-    "target": "esnext"
+    "target": "esnext",
+    "resolveJsonModule": true
```
Comment (styfle) [inline packages/now-build-utils/tsconfig.json]: I've had weird bugs from enabling this feature so you might want to use `JSON.parse(fs.readFile('./file.json'))` instead or just `require('./file.json')`

See #3315 for more info.

File: test/lib/deployment/failed-page.txt [hunk 7]
```diff
@@ -0,0 +1,130 @@
+<!DOCTYPE html>
```
Comment (styfle) [inline test/lib/deployment/failed-page.txt]: I think you can delete this file. We should probably add it to `.gitignore`

File: packages/now-build-utils/tsconfig.json (hunk 6 above)
Comment (AndyBitz) [reply packages/now-build-utils/tsconfig.json]: Done 👍

File: packages/now-build-utils/src/detect-builders.ts [hunk 8]
```diff
@@ -98,45 +95,63 @@ function detectFrontBuilder(
...
 
-  if (outputDirectory) {
-    config.outputDirectory = outputDirectory;
+  if (pkg) {
+    const deps: PackageJson['dependencies'] = {
+      ...pkg.dependencies,
+      ...pkg.devDependencies,
+    };
+
+    if (deps['next']) {
```
Comment (styfle) [inline packages/now-build-utils/src/detect-builders.ts:116]: Why is `next` special cased here?

File: packages/now-build-utils/src/detect-builders.ts [hunk 9]
```diff
@@ -368,14 +381,11 @@ export async function detectBuilders(
...
+  const { outputDirectory, buildCommand, framework } = projectSettings;
 
-  if (detectorResult && detectorResult.buildCommand) {
-    const frontendBuilder = detectFrontBuilder(
-      detectorResult,
-      builders,
-      files,
-      options
-    );
+  if ((pkg && hasBuildScript(pkg)) || buildCommand || framework) {
```
Comment (styfle) [inline packages/now-build-utils/src/detect-builders.ts]:
```suggestion
  if (hasBuildScript(pkg) || buildCommand || framework) {
```
This can be simplified here

File: packages/now-build-utils/src/detect-builders.ts [hunk 10]
```diff
@@ -396,35 +406,53 @@ export async function detectBuilders(
...
+    if (outputDirectory && hasDirectory(outputDirectory, files)) {
+      builders.push({
+        use: '@now/static',
+        src: `${outputDirectory}/**/*`,
+        config: {
+          zeroConfig: true,
+          outputDirectory,
+        },
+      });
+    } else if (hasDirectory('public', files)) {
```
Comment (styfle) [inline packages/now-build-utils/src/detect-builders.ts]: I think you can get rid of this `else if` by setting `outputDirectory` above.

```js
outputDirectory = outputDirectory || 'public';
```

File: packages/now-build-utils/src/detect-builders.ts (hunk 8 above)
Comment (AndyBitz) [reply packages/now-build-utils/src/detect-builders.ts:116]: @styfle because it uses the `@now/next` builder, for now

File: packages/now-build-utils/test/unit.builds-and-routes-detector.test.js [hunk 11]
```diff
@@ -0,0 +1,1161 @@
+const { detectBuilders, detectRoutes } = require('../dist');
```
Comment (styfle) [inline packages/now-build-utils/test/unit.builds-and-routes-detector.test.js]: Can this file use typescript now that we have ts jest?

`unit.builds-and-routes-detector.test.ts`

File: packages/now-cli/package.json [hunk 12]
```diff
@@ -15,7 +15,7 @@
     "test-unit": "nyc ava test/*unit.js --serial --fail-fast --verbose",
     "test-integration": "ava test/integration.js --serial --fail-fast",
     "test-integration-v1": "ava test/integration-v1.js --serial --fail-fast",
-    "test-integration-now-dev": "ava test/dev/integration.js --serial --fail-fast --verbose",
+    "test-integration-now-dev": "NODE_ENV=test ava test/dev/integration.js --serial --fail-fast --verbose",
```
Comment (styfle) [inline packages/now-cli/package.json]: Where is `NODE_ENV` used?

File: packages/now-cli/src/util/dev/server.ts [hunk 13]
```diff
@@ -545,18 +526,13 @@ export default class DevServer {
...
-        detectorResult,
-        {
-          tag: getDistTag(cliVersion) === 'canary' ? 'canary' : 'latest',
-          functions: config.functions,
-        }
-      );
+      const { builders, warnings, errors } = await detectBuilders(files, pkg, {
+        tag: getDistTag(cliVersion) === 'canary' ? 'canary' : 'latest',
+        functions: config.functions,
+        ...(projectSettings ? { projectSettings } : {}),
```
Comment (styfle) [inline packages/now-cli/src/util/dev/server.ts:534]: This is an interesting technique. Are you doing this to avoid setting `projectSettings : undefined`?

File: packages/now-cli/test/dev/integration.js [hunk 14]
```diff
@@ -624,9 +645,7 @@ test('[now dev] 06-gridsome', async t => {
     const response = await fetch(`http://localhost:${port}`);
 
     validateResponseHeaders(t, response);
-
-    const body = await response.text();
-    t.regex(body, /<div id="app"><\/div>/gm);
```
Comment (styfle) [inline packages/now-cli/test/dev/integration.js:629]: Any reason you're not checking the body text anymore?

File: packages/now-cli/test/dev/integration.js [hunk 15]
```diff
@@ -651,8 +670,9 @@ test(
 
     validateResponseHeaders(t, response);
 
-    const body = await response.text();
-    t.regex(body, /Hugo on ZEIT Now/gm);
+    // const body = await response.text();
+    // t.regex(body, /Hugo on ZEIT Now/gm);
```
Comment (styfle) [inline packages/now-cli/test/dev/integration.js:674]: Any reason you're not checking the body text anymore?

File: packages/now-cli/src/util/dev/server.ts (hunk 13 above)
Comment (AndyBitz) [reply packages/now-cli/src/util/dev/server.ts:534]: yes

File: packages/now-cli/test/dev/integration.js (hunk 14 above)
Comment (AndyBitz) [reply packages/now-cli/test/dev/integration.js:629]: Yup, we can change this check back in the future, but for now I'd like to keep those two like this since the detectors changed again, which would cause more failing tests when something gets published again

File: packages/now-cli/test/dev/integration.js (hunk 15 above)
Comment (AndyBitz) [reply packages/now-cli/test/dev/integration.js:674]: same as above

File: packages/now-static-build/src/index.ts [hunk 16]
```diff
@@ -218,12 +216,24 @@ export async function build({
...
+    config.devCommand || (framework && framework.devCommand);
+  const buildCommand: string | undefined =
+    config.buildCommand || (framework && framework.buildCommand);
+
+  console.log({
+    framework,
+    buildCommand,
+    devCommand,
+    config,
+  });
```
Comment (styfle) [inline packages/now-static-build/src/index.ts]: Should this `console.log()` be removed?

File: packages/now-cli/test/dev/integration.js (hunk 14 above)
Comment (styfle) [reply packages/now-cli/test/dev/integration.js:629]: I almost forgot that `now dev` tests still depend on published builders 😖

File: packages/now-build-utils/src/detect-builders.ts [hunk 17]
```diff
@@ -33,8 +33,13 @@ function getApiBuilders({ tag }: Pick<Options, 'tag'> = {}): Builder[] {
   ];
 }
 
-function hasDirectory(fileName: string, files: string[]) {
-  return files.some(name => name.startsWith(`${fileName}/`));
+function hasDirectory(name: string, files: string[]) {
+  return files.some(file => file.startsWith(`${name}/`));
+}
+
+function hasBuildScript(pkg: PackageJson | undefined) {
```
Comment (styfle) [inline packages/now-build-utils/src/detect-builders.ts]:
```suggestion
function hasBuildScript(pkg: PackageJson | null | undefined) {
```

This will fix the build error

File: packages/now-build-utils/test/unit.builds-and-routes-detector.test.js (hunk 11 above)
Comment (AndyBitz) [reply packages/now-build-utils/test/unit.builds-and-routes-detector.test.js]: 👍 converted them, just had to add some casting

File: packages/now-cli/package.json (hunk 12 above)
Comment (AndyBitz) [reply packages/now-cli/package.json]: Not anymore, removed it again

File: packages/now-build-utils/src/detect-builders.ts (hunk 17 above)
Comment (AndyBitz) [reply packages/now-build-utils/src/detect-builders.ts]: Ah, already added it :D

File: packages/frameworks/index.d.ts (hunk 3 above)
Comment (styfle) [reply packages/frameworks/index.d.ts:16]: > at least it looks like it's using the repo url for the logo data.

Can you explain more? What is using the repo url?

Comment (styfle) [reply packages/frameworks/index.d.ts:16]: I see now, its used in an API that front consumes.

**Review by styfle — APPROVED** [review]
Thanks, great work! 🎉
